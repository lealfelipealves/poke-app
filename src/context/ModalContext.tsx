import React, { createContext, useState, useContext } from "react";

type ModalContextProps = {
  children: React.ReactNode;
};

type ModalContextProviderProps = {
  isOpen?: boolean;
  setIsOpen: (open: boolean) => void;
};

export const ModalContext = createContext<ModalContextProviderProps>(
  {} as ModalContextProviderProps
);
ModalContext.displayName = "Modal Context";

export const ModalProvider = ({ children }: ModalContextProps) => {
  let [isOpen, setIsOpen] = useState(false);

  return (
    <ModalContext.Provider value={{ isOpen, setIsOpen }}>
      {children}
    </ModalContext.Provider>
  );
};

export const useModal = () => {
  const context = useContext(ModalContext);
  if (!context) {
    throw new Error(
      "Para usar o useModal, é obrigatório o usuário do Provider"
    );
  }
  return context;
};
