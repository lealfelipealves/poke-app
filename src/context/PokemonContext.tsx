import React, { createContext, useState, useMemo, useContext, useEffect } from 'react';
import { Pokemon, Regions } from 'pokenode-ts'
import { usePokemonInfiniteQuery } from '@/hooks/usePokemonInfiniteQuery';
import { NamedAPIResourceWithId } from '@/types';

type PokemonContextProps = {
  children: React.ReactNode
}

type PokemonContextProviderProps = {
  searchTerm?: string
  setSearchTerm: (searchTerm: string) => void,
  filteredData?: NamedAPIResourceWithId[],
  setFilteredData: (pokemonList: NamedAPIResourceWithId[]) => void,
  pokemon?: Pokemon
  setPokemon: (pokemon: Pokemon) => void,
  regionSelected?: Regions
  setRegionSelected: (region?: Regions) => void
}

export const PokemonContext = createContext<PokemonContextProviderProps>(
  {} as PokemonContextProviderProps
)
PokemonContext.displayName = 'Pokemon Context'

export const PokemonProvider = ({ children }: PokemonContextProps) => {
  const { data } = usePokemonInfiniteQuery();
  
  const [searchTerm, setSearchTerm] = useState("");
  const [filteredData, setFilteredData] = useState<NamedAPIResourceWithId[]>([]);
  const [pokemon, setPokemon] = useState<Pokemon>();  
  const [regionSelected, setRegionSelected] = useState<Regions | undefined>();

  const filterData = () => {
    const filtered = data?.pages.flatMap((page) =>
      page.results.filter((pokemon) => 
        pokemon.name.toLowerCase().includes(searchTerm.toLowerCase())
      )
    ).map((pokemon) => pokemon)
    setFilteredData(filtered ?? []);
  };
  
  useEffect(() => {
    filterData();
  }, [data, searchTerm]);

  const pokemonValue = useMemo(() => ({
    searchTerm, setSearchTerm,
    pokemon, setPokemon,
    filteredData, setFilteredData,
    regionSelected, setRegionSelected
  }), [
    searchTerm, setSearchTerm,
    pokemon, setPokemon,
    filteredData, setFilteredData,
    regionSelected, setRegionSelected
  ]);

  return (
    <PokemonContext.Provider value={pokemonValue}>
      {children}
    </PokemonContext.Provider>
  );
};

export const usePokemon = () => {
  const context = useContext(PokemonContext);
  if (!context) {
    throw new Error(
      'Para usar o usePokemon, é obrigatório o usuário do Provider'
    )
  }
  return context
}
