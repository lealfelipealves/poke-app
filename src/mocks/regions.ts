import { NamedAPIResourceListWithId } from "@/types";

export const regions: NamedAPIResourceListWithId = {
  count: 10,
  next: null,
  previous: null,
  results: [
    {
      id: 1,
      name: "kanto",
      url: "https://pokeapi.co/api/v2/region/1/",
    },
    {
      id: 2,
      name: "johto",
      url: "https://pokeapi.co/api/v2/region/2/",
    },
    {
      id: 3,
      name: "hoenn",
      url: "https://pokeapi.co/api/v2/region/3/",
    },
    {
      id: 4,
      name: "sinnoh",
      url: "https://pokeapi.co/api/v2/region/4/",
    },
    {
      id: 5,
      name: "unova",
      url: "https://pokeapi.co/api/v2/region/5/",
    },
    {
      id: 6,
      name: "kalos",
      url: "https://pokeapi.co/api/v2/region/6/",
    },
    {
      id: 7,
      name: "alola",
      url: "https://pokeapi.co/api/v2/region/7/",
    },
    {
      id: 8,
      name: "galar",
      url: "https://pokeapi.co/api/v2/region/8/",
    },
    {
      id: 9,
      name: "hisui",
      url: "https://pokeapi.co/api/v2/region/9/",
    },
    {
      id: 10,
      name: "paldea",
      url: "https://pokeapi.co/api/v2/region/10/",
    },
  ],
};
