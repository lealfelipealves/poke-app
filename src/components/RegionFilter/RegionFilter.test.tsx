import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import { RegionFilter } from "./index";
import { queryClient } from "@/services/queryClient";
import { data } from "@/mocks/regions";
import { QueryClientProvider } from "@tanstack/react-query";
import { ModalContext } from "@/context/ModalContext";
import { PokemonContext } from "@/context/PokemonContext";
import { bulbasaur } from "@/mocks/bulbasaur";
import React from "react";
import { useQuery } from '@tanstack/react-query';

jest.mock('@tanstack/react-query');

describe("RegionFilter", () => {
  it("renders correctly", async () => {
    const isOpen = true;
    const pokemon = bulbasaur;
    const setIsOpen = jest.fn();

    queryClient.setQueryData(["regions"], data);

    const Wrapper = ({ children }: any) => {
      return (
        <QueryClientProvider client={queryClient}>
          <PokemonContext.Provider value={{ pokemon }}>
            <ModalContext.Provider value={{ isOpen, setIsOpen }}>
              {children}
            </ModalContext.Provider>
          </PokemonContext.Provider>
        </QueryClientProvider>
      );
    };

    const setStateMock = jest.fn();
    const useStateMock: any = (useState: any) => [useState, setStateMock];
    jest.spyOn(React, "useState").mockImplementation(useStateMock);

    useQuery.mockReturnValue({
      isLoading: false,
      data,
      error: false,
    });

    const { container, debug, getByText  } = render(<RegionFilter />, { wrapper: Wrapper });
    expect(container).toMatchSnapshot();

    // const regionButton = await screen.getByRole("button", {
    //   name: /kanto/i,
    // });

    // fireEvent.click(regionButton);

    // expect(regionButton).toHaveClass('bg-red-500 text-white'); // verifique se o bot√£o foi selecionado
    // // expect(setStateMock).toEqual('kanto');
    // expect(setStateMock).toBeCalled();
    

      // const setRegionSelected = jest.fn();


    debug()

  });
});
