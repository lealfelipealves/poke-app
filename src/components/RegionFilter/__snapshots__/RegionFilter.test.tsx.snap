// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegionFilter renders correctly 1`] = `<div />`;
